### PR TITLE
Add classifier_artifacts parameter to java_export to allow users to publish extra artifacts

### PR DESCRIPTION
**Pull request by thirtyseven**
This PR addresses the remaining issues in #805 which got pretty close to being mergeable. I also renamed "extra_artifacts" to "classifier_artifacts" to stick closer to Maven terminology. Also, since javadocs and source jars are special cases of classifier artifacts, I refactored them to use the same code path.

### CODE REVIEW

File: private/rules/java_export.bzl [hunk 1]
```diff
@@ -235,13 +239,16 @@ def maven_export(
         testonly = testonly,
     )
 
+    classifier_artifacts = dict(classifier_artifacts)  # unfreeze
+    classifier_artifacts.setdefault("sources", ":%s-maven-source" % name)
+    classifier_artifacts.setdefault("javadoc", docs_jar)
```
Comment (shs96c) [inline private/rules/java_export.bzl:244]: Neither `sources` not `javadoc` are mandatory in some cases (eg. when publishing a BOM), so please check that these values are set before adding them to the `dict`.

File: private/rules/maven_publish.bzl [hunk 2]
```diff
@@ -24,9 +25,12 @@ def _maven_publish_impl(ctx):
...
     # Expand maven coordinates for any variables to be replaced.
     coordinates = ctx.expand_make_variables("coordinates", ctx.attr.coordinates, {})
-    artifacts_short_path = ctx.file.artifact_jar.short_path if ctx.file.artifact_jar else ""
-    source_short_path = ctx.file.source_jar.short_path if ctx.file.source_jar else ""
-    javadocs_short_path = ctx.file.javadocs.short_path if ctx.file.javadocs else ""
+    artifacts_short_path = ctx.file.artifact.short_path if ctx.file.artifact else ""
+
+    classifier_artifacts_dict = {}
+    for target, classifier in ctx.attr.classifier_artifacts.items():
+        file = target.files.to_list()[0]
```
Comment (shs96c) [inline private/rules/maven_publish.bzl:32]: Hmmm... this makes me nervous. Could we check that there's only one file and print a warning if there's more than one? I've a vague memory that (maybe) the kotlin rules have more than one output sometimes....

**Review by shs96c — CHANGES_REQUESTED** [review]
Looks good, but a couple of minor things.

File: private/rules/maven_publish.bzl (hunk 2 above)
Comment (thirtyseven) [reply private/rules/maven_publish.bzl]: Added some checks here.

**Review by shs96c — APPROVED** [review]

**Review by shs96c — APPROVED** [review]
Alright! Let's see what happens when this lands :) LGTM

**Comment by shs96c** [comment]
@thirtyseven, I hate to ask for more after the delay in getting this review finished, but could you please rebase?

**Comment by thirtyseven** [comment]
> @thirtyseven, I hate to ask for more after the delay in getting this review finished, but could you please rebase?

Done

**Pull request merged**